src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -412,6 +412,10 @@ private function processAnonymousServices(\DOMDocument $xml, $file, $defaults)
 
         // anonymous services "in the wild"
         if (false !== $nodes = $xpath->query('//container:services/container:service[not(@id)]')) {
+            if ($nodes->length) {
+                @trigger_error(sprintf('Top-level anonymous services are deprecated since Symfony 3.4, the "id" attribute will be required in version 4.0 in %s.', $file), E_USER_DEPRECATED);
```

Comment:
maybe we should include the class in the deprecation message to make it easier to find the place